src/net/server/handlers/message_stream_layer/service.rs

```diff
@@ -0,0 +1,134 @@
+use super::future::ResponseFuture;
+use crate::net::server::MpcServerError;
+use axum::http::{Request, Response};
+use std::task::{ready, Context, Poll};
+use tokio::sync::mpsc;
+use tokio_util::sync::PollSender;
+use tower::Service;
+
+#[derive(Debug)]
+pub enum SenderStatus<T> {
+    NotReady(PollSender<T>),
+    Ready(PollSender<T>),
+    Err,
+    Spent,
+}
+
+// manually implement Clone to get around requirements that `T` must be [`Clone`]
+impl<T> Clone for SenderStatus<T> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::NotReady(sender) => Self::NotReady(sender.clone()),
+            Self::Ready(sender) => Self::Ready(sender.clone()),
+            Self::Err => Self::Err,
+            Self::Spent => Self::Spent,
+        }
+    }
+}
+
+pub struct ReservedPermit<T>(PollSender<T>);
+impl<T: Send + 'static> ReservedPermit<T> {
+    pub fn send(&mut self, item: T) -> Result<(), MpcServerError> {
+        Ok(self.0.send_item(item)?)
+    }
+}
+
+/// `poll_ready` will first grab a permit via the `PollSender`. Once ready, the sender is injected
+/// into the Request as an Extension. This can only be done once
+#[derive(Debug)]
+pub struct MessageStream<S, T> {
+    inner: S,
+    sender: SenderStatus<T>,
+}
+
+impl<S, T: Send + 'static> MessageStream<S, T> {
+    pub fn new(inner: S, sender: mpsc::Sender<T>) -> Self {
+        Self {
+            inner,
+            sender: SenderStatus::NotReady(PollSender::new(sender)),
+        }
+    }
+}
+
+/// Within `poll_ready`, first grabs a permit for the [`PollSender`]. Then, pass the [`PollSender`]
+/// into the request as an [`Extension`] on the `call`.
+///
+/// Leaves the [`Service`]'s `Error` type unspecified, so that it can be [`Infallible`] within the
+/// context of Axum. That additionally means never returning an error, instead creating a response
+/// with an error [`StatusCode`]
+impl<S, ReqBody, ResBody, T> Service<Request<ReqBody>> for MessageStream<S, T>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Default,
+    T: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match &mut self.sender {
+            SenderStatus::NotReady(sender) => {
+                if ready!(sender.poll_reserve(cx)).is_ok() {
+                    // swap `NotReady` for `Ready`
+                    if let SenderStatus::NotReady(sender) =
+                        std::mem::replace(&mut self.sender, SenderStatus::Spent)
+                    {
+                        self.sender = SenderStatus::Ready(sender);
```

Comment:
I don't quite get why we need to perform two step change `NotReady` -> `Spent` -> `Ready` here. Can we just go straight from `NotReady` to `Ready`?

Reply:
code is obsolete